### ISSUE TITLE
i386 menus full of x86-64 entries

### ISSUE THREAD

**Issue by rvalles** [netbootxyz/netboot.xyz#588]
Tried to netboot an old pc (slot-A Athlon), and it took trying 20 Linux-based images or so to boot one that worked and didn't give me a x86-64 required error. Needless to mention this is very frustrating.

In many cases (e.g. Gentoo, from Install Linux menu), a i386 version does definitely exist. But the wrong one is downloaded.

Menus should only list entries that work on the current architecture.

I found Debian installer works, which let me proceed with my plans, but only after wasting a lot of time.

**Comment by thelamer** [comment]
Yeah, this is partially my fault I have given zero thought about 32 bit support as I have been adding piles of live images and other tools. 

My personal stance on it is we should fully drop 32 bit support for x86, you can literally pull core series stuff out of the dumpster now. We would never be loop testing it properly and it would always be a pain point. 

But, this is not my project. I cannot speak for the whole community.

**Comment by antonym** [comment]
I think most of the larger distros are slowly starting to drop 32-bit support and we try not to keep EOL versions on the menu.  So over time we probably will phase out the 32-bit support anyways on all the menu Linux menus.

It might make sense to have a smaller 32-bit menu for things like DOS, a few base 32-Bit Linux distros, or a few other utilities that might at least enable hardware hackers that like to play on old equipment but I am unsure of the demand and usually people hacking on older equipment tend to have their own setup anyways.

**Comment by rvalles** [comment]
Indeed, having separate menus per architecture seems to me as the sane thing to do. I understand there's ongoing work on arm; i386 should probably just be treated as its own separate thing, nothing to do with amd64, to avoid the situation described above. Historical versions could go into their own category if there's a desire to maintain them (I honestly don't care about those, but some might). There's a lot of systems that do actually still support i386, but because the menu are shared and not enough care is put into keeping i386 working, it ends up just booting the wrong architecture with the associated error message and frustration.

You're right, I do have my own netboot setup (out of necessity), but there's some value in having a community-built library of boot options. It's the same deal no matter the architecture. Netboot.xyz rocks, but it would be even better if it moved past its focus on amd64.

**Comment by BoKKeR** [comment]
I thought the Architecture selector in the menu was to switch between different architecture images but I am guessing this feature was abandoned.

It is very frustrating when even the utilities under i386 are all x86-64. I feel that it is not necessary for them to be x86-64. To name a few x86-64 only: 
Clonezilla
Gparted
Caine
Kaspersky rescue disc
Rescatux

**Comment by thelamer** [comment]
The last real (emphasis on real not some random embedded thing that doesn't even have network boot support) x86 CPU series that was 32 bit only was the Pentium 4. So if people are shouting from the rooftops to support their hardware I am not listening. 

If the angle is you have some legacy 32 bit software you want to run any proper distro will include architecture flags to pull in from alt repos while running a 64 bit base is.

Pull the band aid off, 32 bit x86 in the computer and server space is dead, anyone claiming it is not is using some ancient Thinkpad with coreboot, and those hardcore users can just download their own installation medium. Refactoring the code and testing of 32 bit OSs is not worth it for such a small obscure subset of users.

**Comment by BoKKeR** [comment]
If that's the stand then the i386 menu should be removed.

**Comment by antonym** [comment]
Yes, we'll look at getting the 32-bit code removed.

**Comment by deep-pipeline** [comment]
Look, I get that it may seems like you're discussing the value of support for stuff which is very out of date and which you believe you could dumpster dive for hardware.. but actually, it seems to me there is likely to be rather a lot of such equipment (just not in parts of the world you may have direct experience of and not doing mainstream activity), that this equipment is perfectly capable of doing some basic functions (particularly in an educational setting), it's got plenty of embedded energy in its manufacture and it works (if people can conveniently get it booting up) and it's really in the linux-waste-not ethos to keep that stuff operating. All of which makes me think it would be a kind and a broad-minded thing to retain a menu item which has within it some support for old/obscure (i386) support even if you don't actively maintain that menu line..

**Comment by thelamer** [comment]
That is completely fine, but there is no reason you need to boot some super obsolete hardware off the internet. 
Just download what you want to run on it and install it, or better yet setup your own limited PXE menus that will actually boot stuff given the extreme memory constraints. 

People making websites don't make sure their stuff still works on IE6, but that doesn't mean you cannot still make html for it.

**Comment by samip5** [comment]
> The last real (emphasis on real not some random embedded thing that doesn't even have network boot support) x86 CPU series that was 32 bit only was the Pentium 4. So if people are shouting from the rooftops to support their hardware I am not listening.

Actually, I think that even Pentium 4 did in fact support 64-bit operation, at least the one I had did so it needs to be quite ancient to not support it.

**Comment by vchrizz** [comment]
I understand if for some users there is no interest to support 32bit / i386. But I also understand, that there are still some users like me, @rvalles , @BoKKeR who seem to still would like to use 32bit / x86 systems. Nevertheless, as @BoKKeR already correctly stated, there are still enough Distros / Tools available which still support 32bit (sidenote, from todays point, Microsoft still supports 32bit on Windows10).
The architecture menu selection also fooled me into thinking, I could for example just pxe-boot gparted for i386 without the hassle of setting up an own environment for it.
That was the primary reason I am excited for netboot.xyz because I would NOT NEED to set up an own environment and just use this.
I dont want to rant here, just ask if there is interest in keeping the architecture menu selection for i386?
If so, I would offer to look into and hopefully fix the issue to have related i386 versions of Distros/Tools working.

**Comment by antonym** [comment]
Are there particular OS distros you are needing?  Part of the problem is just tracking and maintaining all of the ones that still support 32-bit, it adds a lot of additional maintenance and makes the code more complex.  A lot of operating systems have their own way of naming the architecture mirror structure which deviates a lot.  Every one of those options has to be tracked down and tested.

I try to keep the more popular options updated and tested, there's always going to be obscure hardware and operating systems that are out there but it's really hard to support every option with a limited amount of resources.

That said, PRs are always welcome, if there's something in particular you want to add or fix.

**Comment by vchrizz** [comment]
Well, you are right, thats why I would offer to fix the issue here, before I start to setup an own environment.
The maintenance would be the same, if for an own environment or for netboot.xyz where it would be useful also for other users.
How about changing the architecture selection into a pre-selection menu or (if possible) have some architecture auto-detect feature before the menu is shown?
I understand the hassle of going through every option and check for available 32bit support, then again what if some options do not have or drop support of 32bit? So I would ask if some pre-selection/autodetect upfront would make sense?
This way 32bit software/images could be added under x86 and 64bit under x86_64 (and other architectures also respectively).
If something changes, the menu-list could be simple updated by adding/removing software/images.

**Comment by pataquets** [comment]
Just adding my voice in favour of the widest possible choice. Debian and Arch are still pretty widely and well supported distros for 32 bit use. These are useful for using old hardware, which helps on lots of scenarios (not wasting resources by sending them to junk, education, low requirements tasks, etc)
As long as somebody helps maintaining them, I would not withdraw the option to be useful for everyone.
Just my 2c.

**Comment by danboid** [comment]
I'm another person who still uses netboot with 32 bit only x86 machines, mainly netbooks.

You poo poo the x86 but there are still plenty of people actively using and creating new software and hardware for the C64 and Amiga, which make a P4 look like a supercomputer. The RaspberryPi is another example of a 32 bit platform still in regular use, its default distro still being 32 bit. There are millions of supposedly outdated x86 machines still in use that are still perfectly good for certain tasks eg SSH / RDP consoles, playing emulators etc.

Please retain the Debian, Alpine and Slackware x86 installer options as these still work and in addition I would like to see alt rescue added, if I submit a PR maybe?

https://en.altlinux.org/Rescue

Unlike systemrescuecd, alt still create i586 isos. alt rescue is very similar to srcd.

**Comment by antonym** [comment]
What I thought about doing was splitting out the 32-bit arch into it's own sub menu like I did with arm and start with a blank slate with a few of the still supported and popular options that would simplify some of the code paths.

But with mainly myself running the project, I'm more focused on a lot of the heavier usage of new platforms and ongoing updates with more modern gear.  I will review any PRs that come through if they make sense to add some of these back in.  I'll work on at least trying to see if we can get a structure in place that can be built on.

No one is poo pooing x86, but the reality is a lot of support has been getting dropped by the larger players over time and it's time consuming testing every option out there to ensure they work.  I have plenty of obscure hardware too but sometimes the easiest path is just to burn a USB or CD to get that machine to install once hassle free.  iPXE itself doesn't boot on everything either which can create another issue.

AltLinux also used memdisk which was unreliable and it stopped working at one point which is why it got dropped. https://github.com/netbootxyz/netboot.xyz/pull/378#issuecomment-565232750

**Comment by danboid** [comment]
Ah yes! I never saw thelamers response somehow. I shall have to try his suggestion to get alt rescue to pxe boot.

I'd also like to see NetBSD added. Was NetBSD ever considered or has there been no demand?

https://wiki.netbsd.org/tutorials/how_to_install__40__boot__41___netbsd_using_pxelinux/

**Comment by antonym** [comment]
Re NetBSD: https://github.com/netbootxyz/netboot.xyz/issues/78

**Comment by antonym** [comment]
I put up some changes up into development today that should split off 32-bit platforms into a different more restricted menu.  So far I have Alpine and Debian there.  I don't have any old hardware to play with at the moment so if anyone wants to load up the development endpoint from the Utility menu, let me know how it works.

https://github.com/netbootxyz/netboot.xyz/pull/971

**Comment by rvalles** [comment]
>So far I have Alpine and Debian there. I don't have any old hardware to play with at the moment so if anyone wants to load up the development endpoint from the Utility menu, let me know how it works.

On i386, I see the same as usual even after selecting the dev endpoint in Utilities menu. I'm guessing it's somehow not yet effective?

**Comment by antonym** [comment]
Ah, found a bug, give it a try in about 30 or so once the build system pushes out a new version:

https://github.com/netbootxyz/netboot.xyz/commit/181b334096f9a71bdbcc5f465907610b4b76ca9b

**Comment by rvalles** [comment]
I'll try and give it another go in a few hours, when I get the chance.

**Comment by rvalles** [comment]
Tested. It does show a separate list appropriate for the architecture, now.

I am convinced this approach (separate menus per architecture) is best.

**Comment by antonym** [comment]
This is rolled out so closing this up.

**Comment by jwmh** [comment]
Just wanted to express my support, appreciation, and serious gratitude for all who are doing this.
My father (who is 76 years old) is deeply committed to buying used computers from our local used-electronics shop, FreeGeek PDX ( https://freegeek.org ).
He insists on keeping & using a 15-year-old desktop, which has had the PSU and HDD replaced a couple times already... Will probly continue until its motherboard dies.
For him, it's about environmental & social justice -- e.g. rare-earth metal mines... Deep commitment for him, to never buy new, and always use the ole' life fully out of everything.
So anyways.
I realised this thread was closed, but wanted to say...
Major, serious kudos & appreciation.
(Every time something crashes, this makes our lives easier to always have the latest distro version / security updates ready to go.. Not have to re-download, reburn the latest. And he loses his usb flash sticks easily, so i'm planning to put netbootxyz on his router (ddwrt) for him, to make it easier when he needs to restart from scratch or just deal w corruption due to whatever hardware glitches.)

Link to Free Geek, if you want to read about them
https://freegeek.org
(Also, they offer support to anyone wishing to startup a local FreeGeek franchise thriftstore in their community... 501c3, mostly run by hundreds of volunteers!)

**Comment by antonym** [comment]
@jwmh You are most welcome!  I think we are all a little tired of having to download ISOs from scratch each time.  Glad to hear it's making your father's and your life a bit easier.

Let us know if there's something you'd like to see added or missing, and we'll try and get it added.  I would like to expand the 32-bit options over time as I run across more that are actively supported still.

**Comment by rvalles** [comment]
> Let us know if there's something you'd like to see added or missing, and we'll try and get it added. I would like to expand the 32-bit options over time as I run across more that are actively supported still.

The currently listed FreeBSD option (which uses ghostbsd I believe?) is actually 64bit-only.

32bit-friendly systems that could be added include gentoo, netbsd and reactos.

There's of course more (haiku, plan9front, aros, helenos...) but I do not know how netboot-friendly these are.

The main remaining issue currently is that the utility menu is not arch-stratified, so most tools in there actually require 64bit.

**Comment by antonym** [comment]
systemrescue i686, shredos i686, archlinux 32-bit added in latest release, utilities also has a dedicated 32-bit menu and I'll add additional options as I can split them and test.
